tests/cache/test_mesi_controller.py

```diff
@@ -1,116 +1,479 @@
+        self.mock_put_snoop.assert_not_called()
+
+        # Check that the L1 cache was sent invalidate message
+        self.mock_l1_message.assert_not_called()
+
+        # Assert state transition: SHARED -> INVALId
+        self.assertEqual(next_state, MESIState.SHARED)
+
+    def test_snoop_shared_bus_rwim(self):
+        """
+        Test handling a snooped bus RWIM operation for a cacheline in SHARED state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.SHARED, BusOp.RWIM, address)
+
+        # Check that the snoop result was called with HIT
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.HIT)
+
+        # Check that the L1 cache was sent invalidate message
+        self.mock_l1_message.assert_called_with(CacheMessage.INVALIDATELINE, address)
+
+        # Assert state transition: SHARED -> INVALID
+        self.assertEqual(next_state, MESIState.INVALID)
+
+    def test_snoop_modified_bus_read(self):
+        """
+        Test handling a snooped bus read operation for a cacheline in MODIFIED state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.MODIFIED, BusOp.READ, address)
+
+        # Check that the snoop result was put with HITM
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.HITM)
+
+        # Check that the L1 cache was sent get line message
+        self.mock_l1_message.assert_called_with(CacheMessage.GETLINE, address)
+
+        # Check that bus operation was called with WRITE
+        self.mock_bus_op.assert_called_with(BusOp.WRITE, address)
+
+        # Assert state transition: MODIFIED -> SHARED
+        self.assertEqual(next_state, MESIState.SHARED)
+
+    def test_snoop_modified_bus_rwim(self):
+        """
+        Test handling a snooped bus RWIM operation for a cacheline in MODIFIED state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.MODIFIED, BusOp.RWIM, address)
+
+        # Check that the snoop result was put with HITM
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.HITM)
+
+        # Check that bus operation was called with WRITE
+        self.mock_bus_op.assert_called_with(BusOp.WRITE, address)
+
+        # In sequence of calls, first get line and then invalidate
+        expected_l1_calls = [
+            unittest.mock.call(CacheMessage.GETLINE, address),
+            unittest.mock.call(CacheMessage.INVALIDATELINE, address)
+        ]
+
+        # Check that the L1 cache was sent getline and invalidate messages
+        self.mock_l1_message.assert_has_calls(expected_l1_calls, any_order=False)
+
+        # Assert state transition: MODIFIED -> INVALID
+        self.assertEqual(next_state, MESIState.INVALID)
+
+    def test_snoop_modified_bus_invalidate(self):
```

Comment:
@nkanderson I've added this test, not entirely sure how we want to handle this case. Mark is probably expecting us to have thought of invalid MESI operations and how we handled these. Not entirely sure how this one should be handled. I have snoop operations set to not change state when observing another LLC doing a write-back.

Reply:
Since these cases represent an invalid state for the cache as a whole, I think if we're intended to handle them in some way, it would probably need to be throwing an error or at least a warning in the MESI controller itself. Then we'd need to test for those messages here.